dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h

```diff
@@ -0,0 +1,211 @@
+      out[y * sample.out_pitch.x + x] = ConvertSat<OutputType>(
+              in[y * sample.in_pitch.x + x] * sample.contrast + sample.brightness);
+    }
+  }
+}
+
+
+template <typename InputType, typename OutputType, int ndims>
+class BrightnessContrastGpu {
+ private:
+  static constexpr size_t spatial_dims = ndims - 1;
+  using BlockDesc = kernels::BlockDesc<spatial_dims>;
+
+  std::vector<SampleDescriptor<InputType, OutputType, ndims>> sample_descriptors_;
+
+ public:
+  BlockSetup<spatial_dims, -1 /* No channel dimension, only spatial */> block_setup_;
+
+
+  KernelRequirements Setup(KernelContext &context, const InListGPU<InputType, ndims> &in,
+                           const std::vector<float> &brightness, const std::vector<float> &contrast,
+                           const std::vector<Roi<spatial_dims>> &rois = {}) {
+    DALI_ENFORCE(rois.empty() || rois.size() == static_cast<size_t>(in.num_samples()),
+                 "Provide ROIs either for all or none input tensors");
+    DALI_ENFORCE([=]() -> bool {
+        for (const auto &roi : rois) {
+          if (!all_coords(roi.hi >= roi.lo))
+            return false;
+        }
+        return true;
+    }(), "One or more regions of interests are invalid");
+    DALI_ENFORCE([=]()->bool {
+      auto ref_shape = in.shape[0][ndims-1];
+      for (int i=0;i<in.num_samples();i++){
+        if (in.shape[i][ndims-1] != ref_shape) {
+          return false;
+        }
+      }
+      return true;
+    }(), "Number of channels for every image in batch must be equal");
+
+    auto adjusted_rois = AdjustRois(rois, in.shape);
+    auto shape = in.shape[0][ndims-1];
+    KernelRequirements req;
+    ScratchpadEstimator se;
+    TensorListShape<spatial_dims> output_shape({RoiToShape(adjusted_rois, shape)});
+    block_setup_.SetupBlocks(output_shape, true);
+    se.add<SampleDescriptor<InputType, OutputType, ndims>>(AllocType::GPU, in.num_samples());
+    se.add<BlockDesc>(AllocType::GPU, block_setup_.Blocks().size());
+    req.output_shapes = {output_shape};
+    req.scratch_sizes = se.sizes;
+    return req;
+  }
+
+
+  void Run(KernelContext &context, const OutListGPU<OutputType, ndims> &out,
+           const InListGPU<InputType, ndims> &in, const std::vector<float> &brightness,
+           const std::vector<float> &contrast, const std::vector<Roi<spatial_dims>> &rois = {}) {
+    auto sample_descs = CreateSampleDescriptors(in, out, brightness, contrast);
+
+    typename decltype(sample_descs)::value_type *samples_gpu;
+    BlockDesc *blocks_gpu;
+
+    std::tie(samples_gpu, blocks_gpu) = context.scratchpad->ToContiguousGPU(
+            context.gpu.stream, sample_descs, block_setup_.Blocks());
+
+    dim3 grid_dim = block_setup_.GridDim();
+    dim3 block_dim = block_setup_.BlockDim();
+    auto stream = context.gpu.stream;
+
+    BrightnessContrastKernel<<<grid_dim, block_dim, 0, stream>>>(samples_gpu, blocks_gpu);
```

Comment:
what is the `0` here

Reply:
From CUDA docs:

> Ns is of type size_tand specifiesthe number of bytes inshared memory that is dynamicallyallocatedper blockfor this call in addition to the statically allocated memory; this dynamicallyallocatedmemory is used by any of thevariables declared as an external array as mentioned in Section B.2.3;Nsis an optional argumentwhich defaults to 0;

Reply:
It's shared memory size. If you want to use custom stream you need to pass all parameters to kernel launch.